packages/loader/container-loader/src/loader.ts

```diff
@@ -412,3 +392,86 @@ export class Loader implements IHostLoader {
 		);
 	}
 }
+
+/**
+ * Loads container and leaves it in a state where it does not process any ops
+ * If sequence number is provided, loads up to this sequence number and stops there, otherwise stops immediately after loading snapshot.
+ * @internal
+ * */
+export async function loadContainerPaused(
+	loader: ILoader,
+	request: IRequest,
+	loadToSequenceNumber?: number,
+) {
+	const container = await loader.resolve({
+		url: request.url,
+		headers: {
+			...Headers,
+			// ensure we do not process any ops.
+			[LoaderHeader.loadMode]: { opsBeforeReturn: undefined, deltaConnection: "none" },
+		},
+	});
+
+	// If we are trying to pause at a specific sequence number, ensure the latest snapshot is not newer than the desired sequence number.
+	if (loadToSequenceNumber !== undefined) {
+		const lastProcessedSequenceNumber = container.deltaManager.initialSequenceNumber;
+		if (lastProcessedSequenceNumber > loadToSequenceNumber) {
+			throw new Error(
+				"Cannot satisfy request to pause the container at the specified sequence number. Most recent snapshot is newer than the specified sequence number.",
+			);
+		}
+	}
+
+	// Force readonly mode - this will ensure we don't receive an error for the lack of join op
+	container.forceReadonly?.(true);
+
+	// We need to setup a listener to stop op processing once we reach the desired sequence number (if specified).
+	const opHandler = () => {
+		if (loadToSequenceNumber === undefined) {
+			// If there is no specified sequence number, pause after the inbound queue is empty.
+			if (container.deltaManager.inbound.length !== 0) {
+				return;
+			}
+		} else {
+			// If there is a specified sequence number, keep processing until we reach it.
+			if (container.deltaManager.lastSequenceNumber < loadToSequenceNumber) {
+				return;
+			}
+		}
+
+		// Pause op processing once we have processed the desired number of ops.
+		void container.deltaManager.inbound.pause();
+		void container.deltaManager.outbound.pause();
+		container.off("op", opHandler);
+	};
+	if (
+		(loadToSequenceNumber === undefined && container.deltaManager.inbound.length === 0) ||
+		container.deltaManager.lastSequenceNumber === loadToSequenceNumber
+	) {
+		// If we have already reached the desired sequence number, call opHandler() to pause immediately.
+		opHandler();
+	} else {
+		// If we have not yet reached the desired sequence number, setup a listener to pause once we reach it.
+		container.on("op", opHandler);
+	}
+
+	container.connect();
```

Comment:
Also probably better to move the connect call below the creation of the promise below - I don't really know if there's any risk of synchronously reaching the target loadToSequenceNumber before the `container.connect()` call completes, but would make it more obviously impossible.

Reply:
It does not matter as the code above it already installed callback to stop op processing once we reach specific op.
The code below does not stop op processing / kills connection, it only blocks exit from this function until we reach some seq#.
That said, we process ops always on clean stack - any kind of such reentrancy is deadly.

Reply:
My point is the code below will wait forever if it misses its chance to observe the relevant event, and the function will never return.  So if no more "op" events fire after the code below runs, this gets stuck.

Reply:
This is exactly what it did before (as block of code was exected after this.connectToDeltaStream() / this.handleDeltaConnectionArg() calls in Container.load()), but it looks correct to me (other than it does not listen for dispose event on a container). I do not see how it can miss anything.

- if loadToSequenceNumber === undefined, then it exits immediately (IF condition)
- if container.deltaManager.lastSequenceNumber >= loadToSequenceNumber then it exits immediately (IF condition).
- Otherwise, it creates a promise to wait for loadToSequenceNumber. Promise constructor will run synchronously and will run lambda synchronously, so there is no chance of missing the event after testing IF condition
    - if it was the case - we would have a ton of code failing in our repo and in Loop, as that's what people do all the time - sequence like that: if not connected, create promise and wait for "connected" event would be wrong.
    - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise confirms that - it has this language: "A function to be executed by the constructor."

Reply:
I see - you're saying you don't care if ops process prior to establishing that promise because the conditional will prevent the promise from being created/awaited if it's too late to observe the `loadToSequenceNumber` being processed.

Given that your approach is to use `deltaConnection: "none"` to prevent op processing until the first `opHandler` is prepared to watch it happen, I still think it would be more clear to complete full setup prior to connect rather than have a mix of pre-connect setup and post-connect setup.  But I see how this avoids the problem now.